File: app/src/main/java/io/github/hiro/lime/hooks/ReadChecker.java
package io.github.hiro.lime.hooks;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Application;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;
import android.os.Bundle;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import java.io.File;
import java.lang.reflect.Method;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import io.github.hiro.lime.hooks.Constants;
import io.github.hiro.lime.LimeModule;
import io.github.hiro.lime.LimeOptions;
import io.github.libxposed.api.XposedInterface;

public class ReadChecker implements IHook {
    private SQLiteDatabase db_line = null;
    private SQLiteDatabase db_contact = null;
    private String currentChatId = null;

    @Override
    public void hook(LimeModule module, ClassLoader classLoader, LimeOptions limeOptions) throws Throwable {

        // 1. 初始化資料庫與攔截封包 (Hook Application.onCreate)
        try {
            Method onCreateMethod = Application.class.getDeclaredMethod("onCreate");
            module.hook(onCreateMethod, new XposedInterface.Hooker() {
                @Override
                // 【修正】回傳改為 Object
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback<?> callback) throws Throwable {
                    // 【修正】執行原本方法並保留回傳值
                    Object result = callback.callOriginal();
                    
                    Application appContext = (Application) callback.getThisObject();
                    if (appContext == null) return result;

                    File dbFileLine = appContext.getDatabasePath("naver_line");
                    File dbFileContact = appContext.getDatabasePath("contact");

                    if (dbFileLine.exists() && dbFileContact.exists()) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                            SQLiteDatabase.OpenParams params = new SQLiteDatabase.OpenParams.Builder()
                                    .addOpenFlags(SQLiteDatabase.OPEN_READWRITE)
                                    .build();

                            db_line = SQLiteDatabase.openDatabase(dbFileLine, params);
                            db_contact = SQLiteDatabase.openDatabase(dbFileContact, params);

                            try {
                                db_line.execSQL("CREATE TABLE IF NOT EXISTS lime_read_watermarks (chat_id TEXT, reader_mid TEXT, max_msg_id INTEGER, PRIMARY KEY(chat_id, reader_mid))");
                            } catch (Exception ignored) {}

                            hookNetwork(module, classLoader);
                        }
                    }
                    return result;
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "ReadChecker (onCreate) 失敗: " + e.getMessage());
        }

        // 2. 獲取目前進入的聊天室 ID (Hook ChatHistoryRequest.getChatId)
        try {
            Class<?> chatHistoryRequestClass = classLoader.loadClass("com.linecorp.line.chat.request.ChatHistoryRequest");
            Method getChatIdMethod = chatHistoryRequestClass.getDeclaredMethod("getChatId");
            
            module.hook(getChatIdMethod, new XposedInterface.Hooker() {
                @Override
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback<?> callback) throws Throwable {
                    Object result = callback.callOriginal();
                    currentChatId = (String) callback.getResult();
                    return result;
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "ReadChecker (getChatId) 失敗: " + e.getMessage());
        }

        // 3. 加入按鈕 (Hook ChatHistoryActivity.onCreate)
        try {
            Class<?> chatHistoryActivityClass = classLoader.loadClass("jp.naver.line.android.activity.chathistory.ChatHistoryActivity");
            Method activityOnCreate = chatHistoryActivityClass.getDeclaredMethod("onCreate", Bundle.class);

            module.hook(activityOnCreate, new XposedInterface.Hooker() {
                @Override
                public Object intercept(@NonNull XposedInterface.BeforeHookCallback<?> callback) throws Throwable {
                    Object result = callback.callOriginal();
                    Activity activity = (Activity) callback.getThisObject();
                    if (currentChatId != null) {
                        addTopButton(activity);
                    }
                    return result;
                }
            });
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "ReadChecker (Activity) 失敗: " + e.getMessage());
        }
    }

    private void hookNetwork(LimeModule module, ClassLoader classLoader) {
        try {
            Class<?> responseClass = classLoader.loadClass(Constants.RESPONSE_HOOK.className);
            for (Method method : responseClass.getDeclaredMethods()) {
                if (method.getName().equals(Constants.RESPONSE_HOOK.methodName)) {
                    module.hook(method, new XposedInterface.Hooker() {
                        @Override
                        public Object intercept(@NonNull XposedInterface.BeforeHookCallback<?> callback) throws Throwable {
                            Object result = callback.callOriginal();
                            
                            Object[] args = callback.getArgs();
                            if (args == null || args.length < 2 || args[1] == null) return result;

                            String paramValue = args[1].toString();
                            if (paramValue.contains("type:NOTIFIED_READ_MESSAGE")) {
                                String[] operations = paramValue.split("Operation\\(");
                                for (String op : operations) {
                                    if (op.contains("type:NOTIFIED_READ_MESSAGE")) {
                                        String readerMid = extractParam(op, "param2:");
                                        String msgIdStr = extractParam(op, "param3:");

                                        if (readerMid != null && msgIdStr != null && db_line != null) {
                                            try {
                                                long msgId = Long.parseLong(msgIdStr);
                                                try (Cursor c = db_line.rawQuery("SELECT chat_id FROM chat_history WHERE server_id = ?", new String[]{msgIdStr})) {
                                                    if (c.moveToFirst()) {
                                                        String exactChatId = c.getString(0);
                                                        db_line.execSQL("INSERT OR REPLACE INTO lime_read_watermarks (chat_id, reader_mid, max_msg_id) VALUES (?, ?, ?)",
                                                                new Object[]{exactChatId, readerMid, msgId});
                                                    }
                                                }
                                            } catch (Exception e) {
                                                module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "ReadChecker Network Error: " + e.getMessage());
                                            }
                                        }
                                    }
                                }
                            }
                            return result;
                        }
                    });
                }
            }
        } catch (Exception e) {
            module.log(XposedInterface.LOG_LEVEL_ERROR, "LIMEs", "ReadChecker (Network) 失敗: " + e.getMessage());
        }
    }

    // 後續 UI 與邏輯保持不變...
    private void addTopButton(Activity activity) {
        final int BUTTON_ID = 95279527;
        ViewGroup layout = activity.findViewById(android.R.id.content);
        if (layout == null || activity.findViewById(BUTTON_ID) != null) return;

        Button btn = new Button(activity);
        btn.setId(BUTTON_ID);
        btn.setText("👀");
        btn.setBackgroundColor(android.graphics.Color.parseColor("#AA000000"));
        btn.setTextColor(android.graphics.Color.WHITE);
        btn.setTextSize(12);

        android.content.SharedPreferences prefs = activity.getSharedPreferences("LimeReadCheckerPrefs", android.content.Context.MODE_PRIVATE);
        int savedX = prefs.getInt("btn_x", 150);
        int savedY = prefs.getInt("btn_y", 180);

        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
        );
        params.setMargins(savedX, savedY, 0, 0);
        btn.setLayoutParams(params);

        btn.setOnClickListener(v -> {
            if (currentChatId == null || db_line == null || db_contact == null) return;
            Toast.makeText(activity, "資料撈取中...", Toast.LENGTH_SHORT).show();
            new Thread(() -> {
                String resultText = buildReadDataString();
                activity.runOnUiThread(() -> {
                    TextView textView = new TextView(activity);
                    textView.setText(resultText);
                    textView.setPadding(40, 40, 40, 40);
                    ScrollView scrollView = new ScrollView(activity);
                    scrollView.addView(textView);
                    new AlertDialog.Builder(activity)
                            .setTitle("已讀名單 (最近 15 則文字)")
                            .setView(scrollView)
                            .setPositiveButton("關閉", null)
                            .show();
                });
            }).start();
        });

        btn.setOnTouchListener(new android.view.View.OnTouchListener() {
            private int initialX, initialY;
            private float initialTouchX, initialTouchY;
            private boolean isMoved = false;
            @Override
            public boolean onTouch(android.view.View v, android.view.MotionEvent event) {
                FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) v.getLayoutParams();
                switch (event.getAction()) {
                    case android.view.MotionEvent.ACTION_DOWN:
                        initialX = lp.leftMargin; initialY = lp.topMargin;
                        initialTouchX = event.getRawX(); initialTouchY = event.getRawY();
                        isMoved = false; return true;
                    case android.view.MotionEvent.ACTION_MOVE:
                        int dx = (int) (event.getRawX() - initialTouchX);
                        int dy = (int) (event.getRawY() - initialTouchY);
                        if (Math.abs(dx) > 10 || Math.abs(dy) > 10) isMoved = true;
                        lp.leftMargin = initialX + dx; lp.topMargin = initialY + dy;
                        v.setLayoutParams(lp); return true;
                    case android.view.MotionEvent.ACTION_UP:
                        if (!isMoved) v.performClick();
                        else prefs.edit().putInt("btn_x", lp.leftMargin).putInt("btn_y", lp.topMargin).apply();
                        return true;
                }
                return false;
            }
        });
        layout.addView(btn);
    }

    private String buildReadDataString() {
        StringBuilder sb = new StringBuilder();
        try {
            Map<String, Long> watermarks = new HashMap<>();
            try (Cursor cWater = db_line.rawQuery("SELECT reader_mid, max_msg_id FROM lime_read_watermarks WHERE chat_id = ?", new String[]{currentChatId})) {
                while (cWater.moveToNext()) { watermarks.put(cWater.getString(0), cWater.getLong(1)); }
            }
            try (Cursor cMsg = db_line.rawQuery("SELECT server_id, content, created_time FROM chat_history WHERE chat_id = ? AND from_mid IS NULL AND type = 1 ORDER BY CAST(server_id AS INTEGER) DESC LIMIT 15", new String[]{currentChatId})) {
                if (cMsg.getCount() == 0) return "最近沒有傳送文字訊息。";
                while (cMsg.moveToNext()) {
                    long serverId = cMsg.getLong(0);
                    sb.append("💬 ").append(cMsg.getString(1)).append("\n🕒 ").append(formatTime(cMsg.getString(2))).append("\n");
                    List<String> readers = new ArrayList<>();
                    for (Map.Entry<String, Long> entry : watermarks.entrySet()) {
                        if (entry.getValue() >= serverId) {
                            String name = getContactName(entry.getKey());
                            readers.add(name != null ? name : "未知名稱");
                        }
                    }
                    if (readers.isEmpty()) sb.append("❌ 尚無人已讀\n\n");
                    else {
                        sb.append("👀 已讀 (").append(readers.size()).append(")：\n");
                        for (String r : readers) sb.append("  - ").append(r).append("\n");
                        sb.append("\n");
                    }
                }
            }
        } catch (Exception e) { sb.append("發生錯誤：").append(e.getMessage()); }
        return sb.toString();
    }

    private String getContactName(String mid) {
        try (Cursor c = db_contact.rawQuery("SELECT profile_name FROM contacts WHERE mid = ?", new String[]{mid})) {
            if (c.moveToFirst()) return c.getString(0);
        } catch (Exception ignored) {}
        return null;
    }

    private String extractParam(String operation, String prefix) {
        try {
            java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(prefix + "([a-zA-Z0-9]+)");
            java.util.regex.Matcher matcher = pattern.matcher(operation);
            if (matcher.find()) return matcher.group(1);
        } catch (Exception ignored) {}
        return null;
    }

    private String formatTime(String epochStr) {
        try {
            return new SimpleDateFormat("MM-dd HH:mm:ss", Locale.getDefault()).format(new Date(Long.parseLong(epochStr)));
        } catch (Exception e) { return "未知時間"; }
    }
}

File: app/src/main/java/io/github/hiro/lime/hooks/ReadChecker.kt
package io.github.hiro.lime.hooks

import android.app.Activity
import android.app.AlertDialog
import android.app.Application
import android.database.sqlite.SQLiteDatabase
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.*
import io.github.hiro.lime.LimeModule
import io.github.hiro.lime.LimeOptions
import java.io.File
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

class ReadChecker : IHook {
    private var dbLine: SQLiteDatabase? = null
    private var dbContact: SQLiteDatabase? = null
    private var currentChatId: String? = null

    override fun hook(module: LimeModule, classLoader: ClassLoader, limeOptions: LimeOptions) {
        // 1. 初始化資料庫連線 (在 Application onCreate)
        runCatching {
            val onCreateMethod = Application::class.java.getDeclaredMethod("onCreate")
            module.hook(onCreateMethod, module.newHookBuilder().setInterceptor { chain ->
                val result = chain.proceed()
                val appContext = chain.thisObject as? Application ?: return@setInterceptor result

                val dbPathLine = appContext.getDatabasePath("naver_line")
                val dbPathContact = appContext.getDatabasePath("contact")

                if (dbPathLine.exists() && dbPathContact.exists() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                    val params = SQLiteDatabase.OpenParams.Builder()
                        .addOpenFlags(SQLiteDatabase.OPEN_READWRITE)
                        .build()
                    
                    runCatching {
                        dbLine = SQLiteDatabase.openDatabase(dbPathLine, params)
                        dbContact = SQLiteDatabase.openDatabase(dbPathContact, params)
                        
                        // 建立自定義的已讀記錄表
                        dbLine?.execSQL("CREATE TABLE IF NOT EXISTS lime_read_watermarks (chat_id TEXT, reader_mid TEXT, max_msg_id INTEGER, PRIMARY KEY(chat_id, reader_mid))")
                        
                        hookNetwork(module, classLoader)
                    }
                }
                result
            }.build())
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "ReadChecker (onCreate) 失敗: ${e.message}") }

        // 2. 攔截當前聊天室 ID (ChatHistoryRequest)
        runCatching {
            val chatHistoryRequestClass = classLoader.loadClass("com.linecorp.line.chat.request.ChatHistoryRequest")
            val getChatIdMethod = chatHistoryRequestClass.getDeclaredMethod("getChatId")
            module.hook(getChatIdMethod, module.newHookBuilder().setInterceptor { chain ->
                val result = chain.proceed()
                currentChatId = result as? String
                result
            }.build())
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "ReadChecker (getChatId) 失敗: ${e.message}") }

        // 3. 注入 UI 懸浮按鈕 (ChatHistoryActivity)
        runCatching {
            val chatHistoryActivityClass = classLoader.loadClass("jp.naver.line.android.activity.chathistory.ChatHistoryActivity")
            val activityOnCreate = chatHistoryActivityClass.getDeclaredMethod("onCreate", Bundle::class.java)
            module.hook(activityOnCreate, module.newHookBuilder().setInterceptor { chain ->
                val result = chain.proceed()
                val activity = chain.thisObject as? Activity
                if (currentChatId != null && activity != null) {
                    addTopButton(activity)
                }
                result
            }.build())
        }.onFailure { e -> module.log(Log.ERROR, "LIMEs", "ReadChecker (Activity) 失敗: ${e.message}") }
    }

    private fun hookNetwork(module: LimeModule, classLoader: ClassLoader) {
        runCatching {
            val responseClass = classLoader.loadClass(Constants.RESPONSE_HOOK.className)
            responseClass.declaredMethods.forEach { method ->
                if (method.name == Constants.RESPONSE_HOOK.methodName) {
                    module.hook(method, module.newHookBuilder().setInterceptor { chain ->
                        val result = chain.proceed()
                        val args = chain.args
                        if (args.size < 2 || args[1] == null) return@setInterceptor result

                        val paramValue = args[1].toString()
                        if (paramValue.contains("type:NOTIFIED_READ_MESSAGE")) {
                            paramValue.split("Operation(").forEach { op ->
                                if (op.contains("type:NOTIFIED_READ_MESSAGE")) {
                                    val readerMid = extractParam(op, "param2:")
                                    val msgIdStr = extractParam(op, "param3:")
                                    
                                    if (readerMid != null && msgIdStr != null) {
                                        dbLine?.let { db ->
                                            db.rawQuery("SELECT chat_id FROM chat_history WHERE server_id = ?", arrayOf(msgIdStr)).use { c ->
                                                if (c.moveToFirst()) {
                                                    val exactChatId = c.getString(0)
                                                    db.execSQL("INSERT OR REPLACE INTO lime_read_watermarks (chat_id, reader_mid, max_msg_id) VALUES (?, ?, ?)",
                                                        arrayOf(exactChatId, readerMid, msgIdStr.toLong()))
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                        result
                    }.build())
                }
            }
        }
    }

    private fun addTopButton(activity: Activity) {
        val buttonId = 95279527
        val layout = activity.findViewById<ViewGroup>(android.R.id.content) ?: return
        if (activity.findViewById<View>(buttonId) != null) return

        val btn = Button(activity).apply {
            id = buttonId
            text = "👀"
            setBackgroundColor(Color.parseColor("#AA000000"))
            setTextColor(Color.WHITE)
            textSize = 12f
        }

        val prefs = activity.getSharedPreferences("LimeReadCheckerPrefs", Context.MODE_PRIVATE)
        val params = FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT).apply {
            setMargins(prefs.getInt("btn_x", 150), prefs.getInt("btn_y", 180), 0, 0)
        }
        btn.layoutParams = params

        btn.setOnClickListener {
            if (currentChatId == null || dbLine == null) return@setOnClickListener
            Toast.makeText(activity, "資料撈取中...", Toast.LENGTH_SHORT).show()
            Thread {
                val resultText = buildReadDataString()
                activity.runOnUiThread {
                    val textView = TextView(activity).apply {
                        text = resultText
                        setPadding(40, 40, 40, 40)
                    }
                    val scrollView = ScrollView(activity).apply { addView(textView) }
                    AlertDialog.Builder(activity)
                        .setTitle("已讀名單 (最近 15 則文字)")
                        .setView(scrollView)
                        .setPositiveButton("關閉", null)
                        .show()
                }
            }.start()
        }

        // 處理按鈕拖拽
        btn.setOnTouchListener(object : View.OnTouchListener {
            private var initialX = 0
            private var initialY = 0
            private var initialTouchX = 0f
            private var initialTouchY = 0f
            private var isMoved = false

            override fun onTouch(v: View, event: MotionEvent): Boolean {
                val lp = v.layoutParams as FrameLayout.LayoutParams
                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        initialX = lp.leftMargin
                        initialY = lp.topMargin
                        initialTouchX = event.rawX
                        initialTouchY = event.rawY
                        isMoved = false
                        return true
                    }
                    MotionEvent.ACTION_MOVE -> {
                        val dx = (event.rawX - initialTouchX).toInt()
                        val dy = (event.rawY - initialTouchY).toInt()
                        if (Math.abs(dx) > 10 || Math.abs(dy) > 10) isMoved = true
                        lp.leftMargin = initialX + dx
                        lp.topMargin = initialY + dy
                        v.layoutParams = lp
                        return true
                    }
                    MotionEvent.ACTION_UP -> {
                        if (!isMoved) v.performClick()
                        else prefs.edit().putInt("btn_x", lp.leftMargin).putInt("btn_y", lp.topMargin).apply()
                        return true
                    }
                }
                return false
            }
        })
        layout.addView(btn)
    }

    private fun buildReadDataString(): String {
        val sb = StringBuilder()
        runCatching {
            val watermarks = mutableMapOf<String, Long>()
            dbLine?.rawQuery("SELECT reader_mid, max_msg_id FROM lime_read_watermarks WHERE chat_id = ?", arrayOf(currentChatId)).use { c ->
                while (c?.moveToNext() == true) watermarks[c.getString(0)] = c.getLong(1)
            }

            dbLine?.rawQuery("SELECT server_id, content, created_time FROM chat_history WHERE chat_id = ? AND from_mid IS NULL AND type = 1 ORDER BY CAST(server_id AS INTEGER) DESC LIMIT 15", arrayOf(currentChatId)).use { c ->
                if (c == null || c.count == 0) return "最近沒有傳送文字訊息。"
                while (c.moveToNext()) {
                    val serverId = c.getLong(0)
                    sb.append("💬 ").append(c.getString(1)).append("\n🕒 ").append(formatTime(c.getString(2))).append("\n")
                    
                    val readers = watermarks.filter { it.value >= serverId }.keys.map { getContactName(it) ?: "未知名稱" }
                    
                    if (readers.isEmpty()) sb.append("❌ 尚無人已讀\n\n")
                    else {
                        sb.append("👀 已讀 (${readers.size})：\n")
                        readers.forEach { sb.append("  - $it\n") }
                        sb.append("\n")
                    }
                }
            }
        }.onFailure { sb.append("發生錯誤：${it.message}") }
        return sb.toString()
    }

    private fun getContactName(mid: String): String? {
        return dbContact?.rawQuery("SELECT profile_name FROM contacts WHERE mid = ?", arrayOf(mid)).use { c ->
            if (c?.moveToFirst() == true) c.getString(0) else null
        }
    }

    private fun extractParam(operation: String, prefix: String): String? {
        val matcher = Pattern.compile("$prefix([a-zA-Z0-9]+)").matcher(operation)
        return if (matcher.find()) matcher.group(1) else null
    }

    private fun formatTime(epochStr: String): String {
        return runCatching {
            SimpleDateFormat("MM-dd HH:mm:ss", Locale.getDefault()).format(Date(epochStr.toLong()))
        }.getOrDefault("未知時間")
    }
}
